test/Container/ConfigFactoryTest.php

```diff
@@ -0,0 +1,130 @@
+<?php
+/**
+ * @see       https://github.com/zendframework/zend-di for the canonical source repository
+ * @copyright Copyright (c) 2017 Zend Technologies USA Inc. (http://www.zend.com)
+ * @license   https://github.com/zendframework/zend-di/blob/master/LICENSE.md New BSD License
+ */
+
+namespace ZendTest\Di\Container;
+
+use PHPUnit\Framework\TestCase;
+use PHPUnit\Framework\Error\Deprecated as DeprecatedError;
+use Psr\Container\ContainerInterface;
+use Zend\Di\Container\ConfigFactory;
+use Zend\Di\ConfigInterface;
+
+/**
+ * @coversDefaultClass Zend\Di\Container\ConfigFactory
+ */
+class ConfigFactoryTest extends TestCase
+{
+    /**
+     * @var \PHPUnit_Framework_MockObject_MockBuilder
+     */
+    private $containerBuilder;
+
+    protected function setUp()
+    {
+        parent::setUp();
+        $this->containerBuilder = $this->getMockBuilder(ContainerInterface::class);
+    }
+
+    protected function tearDown()
+    {
+        DeprecatedError::$enabled = true;
+        $this->containerBuilder = null;
+
+        parent::tearDown();
+    }
+
+    public function testInvokeCreatesConfigInstance()
+    {
+        $container = $this->containerBuilder->getMockForAbstractClass();
+        $container->method('has')->willReturn(false);
+
+        $factory = new ConfigFactory();
+        $this->assertInstanceOf(ConfigInterface::class, $factory($container));
+    }
+
+    /**
+     * The factory must succeed even if the container does not provide "config"
+     */
+    public function testCreateRequestsContainerForConfigServiceGracefully()
+    {
+        $container = $this->containerBuilder->getMockForAbstractClass();
+        $container->expects($this->atLeastOnce())
+            ->method('has')
+            ->with('config')
+            ->willReturn(false);
+
+        $container->expects($this->never())
+            ->method('get')
+            ->with('config');
+
+        $result = (new ConfigFactory())->create($container);
+        $this->assertInstanceOf(ConfigInterface::class, $result);
+    }
+
+    private function createContainerWithConfig($config)
+    {
+        $container = $this->containerBuilder->getMockForAbstractClass();
+        $container->expects($this->atLeastOnce())
+            ->method('has')
+            ->with('config')
+            ->willReturn(true);
+
+        $container->expects($this->atLeastOnce())
+            ->method('get')
+            ->with('config')
+            ->willReturn($config);
+
+        return $container;
+    }
+
+    public function testCreateUsesConfigFromContainer()
+    {
+        $expectedPreference = uniqid('SomePreference');
+        $container = $this->createContainerWithConfig([
+            'dependencies' => [
+                'auto' => [
+                    'preferences' => [
+                        'SomeDependency' => $expectedPreference
+                    ]
+                ]
+            ]
```

Comment:
Note to self: add trailing commas after each array item here and in other test cases in this file.